File: capgemini/batch-1/03-02-24/TwoUnique.java
class TwoUnique{
    public static void main(String[] args){
        int[] arr = {1,8,9,1,8,7,6,3,7,3,6,7};
        int allXor = 0;
        int ans1 = 0, ans2 = 0;
        for(int e : arr) allXor = allXor^e;
        int bitPosition = getFirstSetBit(allXor);
        for(int e : arr){
           if (isIthBitSet(e,bitPosition)){
                ans1 = ans1^e;
           }else{  
                ans2 = ans2^e;
           }
        }
        System.out.println(ans1);
        System.out.println(ans2);
    }
    public static int getFirstSetBit(int n){
        if(n<=0){
            return 0;
        }
        int ans = 1;
        while(n%2 !=1){
            n = n>>1;
            ans+=1;
        }
        return ans;
    }
    public static boolean isIthBitSet(int n, int i){
        n = n>>(i-1);
        return (n%2 == 1);

    }
}

File: capgemini/batch-3/08-02-24/TwoUnique.py
def getFirstSetBit(num):
    ans = 1
    while num%2 != 1:
        num = num>>1
        ans+=1
    return ans
def isIthBitSet(num,i):
    num = num>>(i-1)
    return (num%2 == 1)

def getBothUniqueNumber(arr):
    allXor = 0
    for e in arr: allXor = allXor^e
    bitPos = getFirstSetBit(allXor)
    ans1 = 0
    ans2 = 0
    for e in arr:
        if isIthBitSet(e,bitPos) == True:
            ans1 = ans1^e
        else:
            ans2 = ans2^e
    return [ans1,ans2]



def main():
    arr = [1,2,6,2,1,7,9,7,8,8]
    ans = getBothUniqueNumber(arr)
    print(ans)

main()